EIPS/eip-2315.md

```diff
@@ -2,141 +2,159 @@
+2. Push the current `pc + 3` to the `return stack`.
+3. Set `pc` to `destination`.
 
-Taken together with other required propoposals they provide an _static, complete, safe, and efficient_ control-flow facility.
+The gas cost is _low_.
 
-_Static_.  All possible jumps are known at contract creation time.
+#### `RETURNSUB (0x??)`
 
-_Complete_. Together with [EIP-4200](./eip-4200.md) it provides a minimal set of static control structures for implementing programs -- jumps, conditional jumps, and subroutines.
+Returns control to the caller of a subroutine.
 
-_Safe_.  [EIP-3779](./eip-3779.md) establishes rules for using these control structures such that valid programs will not halt with an exception unless they run out of gas or recursively overflow stack.
+1. Pop the `return stack` to `PC`.
 
-_Efficient_.  Substantial reductions in the costs, and complexity of calling and optimizing simple subroutines.
+The destination address is relative to the current PC. The address is encoded as a two-byte, twos-complement signed integer, stored MSB-first.
 
+The gas cost is _verylow_.
 
-## Specification
+_Notes:_
+* _Values popped off the `return stack` do not need to be validated, since they are alterable only by `RJUMPSUB` and `RETURNSUB`._ 
+* _The description above lays out the semantics of these instructions in terms of a `return stack`.  But the actual state of the `return stack` is not observable by EVM code or consensus-critical to the protocol.  (For example, a node implementer may code `RJUMPSUB` to unobservably push `PC` on the `return stack` rather than `PC + 1`, which is allowed so long as `RETURNSUB` observably returns control to the `PC + 3 location.)_
 
-We are following the [the original specification](http://www.alanturing.net/turing_archive/archive/p/p01/P01-001.html) of subroutines for [Turing's Automatic Computing Engine of 1946](https://doi.org/10.1093/comjnl/20.3.269):
-> We also wish to be able to arrange for the splitting up of operations into subsidiary operations.  This should be done in such a way that once we have written down how an operation is done we can use it as a subsidiary to any other operation.
->  ... 
-> When we wish to start on a subsidiary operation we need only make a note of where we left off the major operation and then apply the first instruction of the subsidiary.  When the subsidiary is over we look up the note [on] a list of these notes in one or more standard size delay lines, (1024) with the most recent last. _[Delay lines were made of mercury-filled crystals!]_
+### Code Validation
 
-That is, to call a subroutine the machine
-  * pushes the address of the next instruction on a `return stack` and
-  * transfers control to the subroutine address.
+If the execution of an instruction would violate a condition, then the execution is in an exceptional halting state. The [Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf) defines five such states.
```

Comment:
```suggestion
If the execution of an instruction would violate a condition, then the execution is in an exceptional halting state. The Yellow Paper defines five such states.
```

Reply:
Is this not a stable link to a normative reference?

Reply:
We probably *should* make an exception for the yellow paper since that is currently the only specification for Ethereum at the moment, though I'm not sure where the *right* place to link to it is.

Reply:
I moved the three links to proper citations in the (controversial) References section, IETF-style.  I used the Yellow Paper URL shown above, which is also linked to from the ethereum/yellowpaper repo, which would be a possible relative link.

Reply:
Moving them to a section that doesn't exist just means your EIP now has two problems instead of one.  😆

Reply:
Yes.  And if I remove the Reference section I will instead have the problem that I am making bald assertions of fact with nothing to back them up.  And if I remove those assertions the proposal will be unmotivated and without rationale.